### PR TITLE
🎨 introduce logger filtering with ENV var ⚠️

### PR DESCRIPTION
**Pull request by matusdrobuliak66**
## What do these changes do?
- We will change the log level to INFO in production. I thought this might be a good approach to interactively filter out some repetitive logs that are not useful and might be polluting the logs.
- 🎨 introduce logger filtering with ENV var `LOG_FILTER_MAPPING`
Example:
```
LOG_FILTER_MAPPING = {
    "uvicorn.access": [
        '"GET / HTTP/1.1" 200',
        '"GET /metrics HTTP/1.1" 200',
    ],
    "gunicorn.access": [
        '"GET /v0/ HTTP/1.1" 200',
        '"GET /v0/health HTTP/1.1" 200',
    ],
}
```

## Dev-ops checklist
 - https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/1029
- [ ] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: packages/service-library/src/servicelib/logging_utils.py
Comment (matusdrobuliak66) [inline packages/service-library/src/servicelib/logging_utils.py:108]: Idea here is that `LOGGER_FILTER_MAPPING` will be taken from the Application settings and passed to the `config_all_loggers` function similar to `log_format_local_dev_enabled`. What do you think about this approach?

Comment (sanderegg) [reply packages/service-library/src/servicelib/logging_utils.py:108]: yes I like it.
If I get it right you will create some layered approach right?

the uvicorn ones for uvicorn-based services, the gunicorn for gunicorn-based ones?
We might also incorporate the NOISY_LOGGERS, which we start having all over the place, and there we could also have some list that can be layered as we have a few we know are especially noisy (aiopika to name a few)?

now I am a tiny little bit concerned about performance of that thing, since we log millions of logs this can have an impact. but we can test, maybe the jaeger can help to find out.

**Review by sanderegg — COMMENTED** [review]
added a comment

File: packages/service-library/src/servicelib/logging_utils.py
Comment (matusdrobuliak66) [reply packages/service-library/src/servicelib/logging_utils.py:110]: Okay, I have changed it to an environment variable in my last commit. 👍 Correct, we can extend this and also add NOISY_LOGGERS as the next step. Regarding performance, we can test it once it's in the master branch. By default, it's an empty dictionary. Performance should be "worsen" only for the loggers that are provided in the ENV var as only for those filtering is happening.

File: .env-devel
Comment (mrnicegyu11) [inline .env-devel:13]: I guess this also needs to go into the `repo.config` files

File: packages/service-library/src/servicelib/logging_utils.py
Comment (mrnicegyu11) [inline packages/service-library/src/servicelib/logging_utils.py:115]: very minor: This log should be info level imho

File: .env-devel
Comment (matusdrobuliak66) [reply .env-devel:13]: Yes it is already prepared there :+1:

File: packages/service-library/src/servicelib/logging_utils.py
Comment (mrnicegyu11) [inline packages/service-library/src/servicelib/logging_utils.py:149]: You may think this is a premature overoptimization but here I would really suggest to go for regexes right away. They are meant for this. You can easily validate if some input string that claims to be a regex is actually a regex. Everyone can lookup regexes on their own if log whitespacing changes or so, and optimize.

Comment (mrnicegyu11) [reply packages/service-library/src/servicelib/logging_utils.py:149]: I mean, match input regexes with the loglines instead of using python's `in`

Comment (mrnicegyu11) [reply packages/service-library/src/servicelib/logging_utils.py:149]: Or globbing would also be nice, in case you somehow dont like regex. I'd prefer regexes over globbing tho.

File: services/agent/src/simcore_service_agent/core/settings.py
Comment (mrnicegyu11) [inline services/agent/src/simcore_service_agent/core/settings.py:28]: I would recommend, since we have pydantic tightly integrated anyway, to use

1.) A saner layout for this dict than the scary, whitespace containing str(`'"GET / HTTP/1.1" 200"'`). I guess you are literally looking if the python logical condition `log-filter-string in logline == True`, which is a quick win. As mentioned before, I would right away go for regexes. Then, if e.g. the number of strings changes or so, we can quickly fix it with general regexes. Regexes in my mind are just the way to go for pattern matching, almost always

2. Dont just use a `dict[any,any]`, but restrict the content of the dict to valid loggers and valid regexes. Pydantic will validate this for you for free then.


Hearty suggestions but I wont break your balls over them.

**Review by mrnicegyu11 — COMMENTED** [review]
I have a subtle change request that I personally find reasonable, but I have formulated as a comment. If you dont want to do it, please bring me a coffee or tell me a compliment, then I might overlook my precautions and approve 😽

File: packages/service-library/src/servicelib/logging_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/logging_utils.py:119]: is this an unexpected condition? if yes I think this should raise not continue silently. Otherwise no one will ever find it out

Comment (sanderegg) [reply packages/service-library/src/servicelib/logging_utils.py:119]: at least an assert would be in order so that this would be catched in tests

File: packages/service-library/src/servicelib/logging_utils.py
Comment (sanderegg) [reply packages/service-library/src/servicelib/logging_utils.py:149]: I do agree that regex might be a better one, at least from usage point of view. although they are maybe slower than straight string matching. And as I said I am wondering a bit on the performance impact of that thing.

File: packages/service-library/tests/test_logging_utils_filter.py
Comment (sanderegg) [inline packages/service-library/tests/test_logging_utils_filter.py:61]: actually you could benchmark here and compare if the logger performance got impeeded.
see [https://github.com/ionelmc/pytest-benchmark](https://github.com/ionelmc/pytest-benchmark) we use it in some places in the repo already.

File: .env-devel
Comment (sanderegg) [inline .env-devel:94]: Minor:
why the changes in this file at all?
is this used in OPS? I thought it was not anymore?

some local computers might not like having suddenly many more logs. just a wonder.

File: services/storage/src/simcore_service_storage/cli.py
Comment (sanderegg) [inline services/storage/src/simcore_service_storage/cli.py:30]: on the line below that is what I meant with the NOISY_LOGGERS.

**Review by sanderegg — APPROVED** [review]
thanks, just a few comments

File: packages/service-library/src/servicelib/logging_utils.py
Comment (matusdrobuliak66) [reply packages/service-library/src/servicelib/logging_utils.py:149]: The `in` operator is generally faster when you need to check if a substring exists within a string and it's simple python approach. Regular expressions are slower than `in` because they offer more flexibility and need to compile and process the pattern to perform the search. I think in this use case using `in` is better option. Also it doesn't prevent us in the future to change it to regex if needed.

File: services/agent/src/simcore_service_agent/core/settings.py
Comment (matusdrobuliak66) [reply services/agent/src/simcore_service_agent/core/settings.py:28]: I do not want to restrict the code content to valid loggers because the purpose of this PR is to provide flexibility. Now, when we switch to INFO logs in production, we have the option to filter certain logs using an environment variable. This is a simple and straightforward approach. As mentioned before, this can be improved in the future if needed. I prefer not to 'overcomplicate' things that haven't been used yet.

File: packages/service-library/src/servicelib/logging_utils.py
Comment (matusdrobuliak66) [reply packages/service-library/src/servicelib/logging_utils.py:149]: https://github.com/ITISFoundation/osparc-simcore/pull/6596#discussion_r1815219159 Yes performance will be observed.

File: .env-devel
Comment (matusdrobuliak66) [reply .env-devel:94]: Which computers? I find it more useful when I start `make up-prod` when I see INFO logs.

File: services/storage/src/simcore_service_storage/cli.py
Comment (matusdrobuliak66) [reply services/storage/src/simcore_service_storage/cli.py:30]: Yes, we can move to similar logic and have an ENV var for those.

File: packages/service-library/src/servicelib/logging_utils.py
Comment (matusdrobuliak66) [reply packages/service-library/src/servicelib/logging_utils.py:119]: No it is expected, as `servicelib` is used across all services and some let say 3rd party loggers might not be used in some of those services.

**Comment by YuryHrytsuk** [comment]
@matusdrobuliak66 this is amazing. I like it! 

I wonder if it further slows down general performance because of logging. It would be nice to see a benchmark but we don't have anything set up yet to compare.

**Review by YuryHrytsuk — COMMENTED** [review]
I think before it is merged, we need a PR in `osparc-config` with all the new required ENV.

**Comment by matusdrobuliak66** [comment]
> @matusdrobuliak66 this is amazing. I like it!
> 
> I wonder if it further slows down general performance because of logging. It would be nice to see a benchmark but we don't have anything set up yet to compare.

When we merge it as is, there will be no performance degradation. If you add a specific logger to filter out messages, performance impact will be minimal, as it only affects that logger (and the logger filter is designed precisely for this purpose). I'd like to conduct some real benchmarking on the master branch, but I’ll need a few approvals for that. 😊

**Comment by matusdrobuliak66** [comment]
> I think before it is merged, we need a PR in `osparc-config` with all the new required ENV.

No PR is needed, as default value is empty string.

**Review by mrnicegyu11 — APPROVED** [review]

**Comment by mrnicegyu11** [comment]
For the record re regex speed: you are right @matusdrobuliak66 https://stackoverflow.com/questions/19911508/python-speed-for-in-vs-regular-expression

File: services/web/server/src/simcore_service_webserver/application_settings.py
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/application_settings.py:119]: NOTE that the `WEBSERVER_LOG_FILTER_MAPPING` (or any other `$(SERVICENAME)_LOG_FILTER_MAPPING`) environment variable is not usable because it has not been defined or exposed in the `docker-compose.yml` file for the relevant services.

Using alias mappings for these variables is outdated and constrained by our current `docker-compose` setup:

1. **Both Variables Exposed:**
   ```yaml
   environment:
     WEBSERVER_LOG_FILTER_MAPPING=${WEBSERVER_LOG_FILTER_MAPPING}
     LOG_FILTER_MAPPING=${LOG_FILTER_MAPPING}
   ```
   In this configuration, `LOG_FILTER_MAPPING` will be ignored in favor of `WEBSERVER_LOG_FILTER_MAPPING`, regardless of its value.

2. **Mapped Differently in `.env` for Each Service:**
   ```yaml
   environment:
     WEBSERVER_LOG_FILTER_MAPPING=${LOG_FILTER_MAPPING}
   ```
   Here, `LOG_FILTER_MAPPING` is mapped directly to `WEBSERVER_LOG_FILTER_MAPPING` in the `.env` file, allowing customization for each service.

3. **Direct One-to-One Mapping:**
   ```yaml
   environment:
     WEBSERVER_LOG_FILTER_MAPPING=${WEBSERVER_LOG_FILTER_MAPPING}
   ```
   This setup maintains a straightforward one-to-one mapping between `WEBSERVER_LOG_FILTER_MAPPING` and its environment variable, without any aliasing.

File: packages/service-library/src/servicelib/logging_utils.py
Comment (pcrespov) [inline packages/service-library/src/servicelib/logging_utils.py:139]: `logging_utils` is becoming very big. What about moving this to `logging_filters.py`? Then rename as well the test etc...

File: packages/service-library/src/servicelib/logging_utils.py
Comment (pcrespov) [inline packages/service-library/src/servicelib/logging_utils.py:140]: annotate inputs

File: services/autoscaling/src/simcore_service_autoscaling/core/settings.py
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/core/settings.py:243]: what about doing some validation for `LOG_FILTER_MAPPING` ?? 
> is a dictionary that maps specific loggers (such as 'uvicorn.access' or 'gunicorn.access') to a list of log message patterns that should be filtered out.

You can define the validator once and reuse it in all settings classes

File: services/autoscaling/src/simcore_service_autoscaling/core/settings.py
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/core/settings.py:244]: use `default_factory=dict` here

File: packages/service-library/src/servicelib/logging_utils.py
Comment (pcrespov) [inline packages/service-library/src/servicelib/logging_utils.py:90]: Document your intent using code ... even if you do not do a hard validation here

e.g. (see notes below as well)

```python

LoggerName: TypeAlias = str
KeywordStr: TypeAlias = str


def foo(..., exclude_keywords_log_filter_args: dict[LoggerName, set[KeywordStr]] )
   ...
```


Let's talk about this offline.

File: packages/service-library/src/servicelib/logging_utils.py
Comment (pcrespov) [reply packages/service-library/src/servicelib/logging_utils.py:149]: what a `GeneralLogFilter` has to do with `filtered_routes`?

You have a list of strings and want to make sure that if the log message include any of these strings, it gets filtered out.


```python
class ExcludeKeywordsLogFilter(logging.Filter)
    def __init__(self, keywords: set[str])
        super().__init__()
        self.excluded_keywords = keywords

    def filter(self, record):
        log_msg = record.getMessage()
        found = any( keyword in log_msg for keyword in self.excluded_keywords )
        return not found
```

Whether to use regex or not, i think this you have to really measure it. My bet is that this method is faster for this particular use case

Comment (pcrespov) [reply packages/service-library/src/servicelib/logging_utils.py:149]: I actually recommend a test (e.g. using pytest-benchmark) that keeps in check the timing with and without filter

**Review by pcrespov — COMMENTED** [review]
I like the idea. I have some comments regarding the implementation. Lots of thoughts  :-)
I am happy to discuss it offline and even help implementing some parts (now or in further iterations)

**Review by pcrespov — APPROVED** [review]

**Review by pcrespov — COMMENTED** [review]
Please consider the comments before merging. Either by taking action in this PR or by creating an issue.

File: services/api-server/src/simcore_service_api_server/core/settings.py
Comment (GitHK) [inline services/api-server/src/simcore_service_api_server/core/settings.py:59]:
```suggestion
        default_factory=dict,
```

File: services/catalog/src/simcore_service_catalog/core/settings.py
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/core/settings.py:65]:
```suggestion
        default_factory=dict,
```

File: services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/core/settings.py
Comment (GitHK) [inline services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/core/settings.py:38]:
```suggestion
        default_factory=dict,
```

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py
Comment (GitHK) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:137]:
```suggestion
        default_factory=dict,
```

File: services/invitations/src/simcore_service_invitations/core/settings.py
Comment (GitHK) [inline services/invitations/src/simcore_service_invitations/core/settings.py:53]:
```suggestion
        default_factory=dict,
```

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/core/settings.py
Comment (GitHK) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/core/settings.py:63]:
```suggestion
        default_factory=dict,
```

File: services/storage/src/simcore_service_storage/settings.py
Comment (GitHK) [inline services/storage/src/simcore_service_storage/settings.py:72]:
```suggestion
        default_factory=dict,
```

**Review by YuryHrytsuk — APPROVED** [review]

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/settings.py:138]: what about this one? it was not changed?

**Review by GitHK — APPROVED** [review]
Thanks. You missed one.

**Pull request merged**
